_posts/2020-11-09-defensive-programing.md

```diff
@@ -0,0 +1,110 @@
+```r
+if (nrow(input_data) == 0) {
+  stop("ERROR: Input data is empty. Did you make a typo?")
+```
+
+In this case, the code is checking if the input file actually has data in it. If the file is empty, it will show the message `ERROR: Input data is empty. Did you make a typo?`. Notice that the statement points out what went wrong and gives an idea of how to solve the problem.
+
+## Some methods of defensive programming
+
+* **Assertions**
+
+Assertions are statements that check if a condition is true or false. If it is, the program does nothing. If the condition is false, the program will print an error message.
+
+Example:
+```r
+numbers = [1.5, 2.3, 0.7, -0.001, 4.4]
+total = 0.0
+for n in numbers:
+    assert n >= 0.0, 'Data should only contain positive values'
+    total += n
+print 'total is:', total
+```
+In this piece of code, the for loop will work only with positive numbers. When it encounters a negative number, it will stop the loop and print the error message `Data should only contain positive values`.
+
+* **Pre-conditions**
+
+Pre-conditions are statements that make sure certain conditions are true before it runs the code. If the conditions are not met, the code doesn’t run, and an error message can be displayed explaining what happened.
+
+Example:
+```r
+public void CreateAppointment(DateTime dateTime)
+{
+    if (dateTime.Date < DateTime.Now.AddDays(1).Date)
+        throw new ArgumentException("Date is too early");
+
+    if (dateTime.Date > DateTime.Now.AddMonths(1).Date)
+        throw new ArgumentException("Date is too late");
+
+    /* Create an appointment */
+}
+```
+Here, the code will only make the appointment if the the given date is correct, not too early or too late.
+
+* **Nulls**
+
+Null is a pointer used in programming for uninitialized, undefined, empty, or meaningless value. Because many programs cannot discriminate null values, they are often the reason for bugs in codes. Therefore, checking your code for nulls when applicable is always a good idea.
+
+Example:
+```r
+public class Controller
+{
+    public Controller(ILogger logger, IEmailGateway gateway)
+    {
+        if (logger == null)
+            throw new ArgumentNullException();
+        if (gateway == null)
+            throw new ArgumentNullException();
+
+        /* ... */
+    }
+
+    public void Process(User user, Order order)
+    {
+        if (user == null)
+            throw new ArgumentNullException();
+
+        /* ... */
+    }
+}
+```
+In this case, the code is checking for `null` values.
+
+## Why is defensive programming important?
+
+Defensive programming is a good practice that allows the user to react to unexpected errors and fix them faster. Let’s say that you create a code that processes extensive tabular data, but you do not use defensive programming. The code works perfectly with one dataset, great! However, when you try to use the same code to process a similar tabular dataset.  
```

Comment:
I'm not sure if you completely finished writing the last sentence in the paragraph "However, when you try..."

Reply:
True! I fixed it.